packages/react/src/components/DesignLibrary.tsx

```
@@ -164,46 +168,73 @@ export const VariantGeneration = (props: VariantGenerationProps) => {
   const [initError, setInitError] = useState<boolean>(false);
   const [Component, setComponent] = useState<DynamicComponent>(null);
 
-  if (!props.importMap) {
-    return <div>No import map found. Please check your import map.</div>;
+  if (!props.importMapImport) {
+    return (
+      <div>
+        No dynamic import map loaded. Please check a dynamic import map function is passed into
+        Design Library
+      </div>
+    );
   }
 
   if (!rendering) {
     return <div>No component found in layout data. Please check your layout data.</div>;
   }
 
   useEffect(() => {
-    const unsubscribe = addComponentPreviewHandler(props.importMap, (error, Component) => {
-      setRenderKey((key) => key + 1);
-
-      if (error) {
-        setInitError(true);
-
-        return;
+    let cancelled = false;
+    // since import map is loaded lazily, we only need to add preview event handler once the import map is loaded
+    // unsubscribe function for useEffect cleanup will also be returned once importMap promise has been resolved or rejected
+    let unsubscribe: (() => void) | undefined = undefined;
+
+    const init = async () => {
+      let importMap: codegen.ImportEntry[] = undefined;
+      try {
+        importMap = (await props.importMapImport()).default;
```

@stasmaxymov could you clarify here: if import map is not resolved, should an error be thrown or and empty importMap passed into event handler?

---

I would throw an error event since the import map resolver was passed intentionally, but something is happened

---

Instead of throwing an error I'm indicating an error state through setInitError, but the details will be logged too.